components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java

```diff
@@ -0,0 +1,414 @@
+            handleActionPropertyResolverClientException(e.getCause());
+            LOG.debug("Error while creating the Action of Action Type: " + actionDTO.getType().getDisplayName() +
+                            " in Tenant Domain: " + IdentityTenantUtil.getTenantDomain(tenantId) +
+                            ". Rolling back created action information, authentication secrets and action properties.");
+            throw ActionManagementExceptionHandler.handleServerException(ErrorMessage.ERROR_WHILE_ADDING_ACTION, e);
+        }
+    }
+
+    @Override
+    public List<ActionDTO> getActionsByActionType(String actionType, Integer tenantId) throws ActionMgtException {
+
+        try {
+            List<ActionDTO> actionDTOS = actionManagementDAO.getActionsByActionType(actionType, tenantId);
+
+            return getActionDTOsWithPopulatedProperties(actionType, actionDTOS, tenantId);
+        } catch (ActionMgtException | ActionPropertyResolverException e) {
+            throw ActionManagementExceptionHandler.handleServerException(
+                    ErrorMessage.ERROR_WHILE_RETRIEVING_ACTIONS_BY_ACTION_TYPE, e);
+        }
+    }
+
+    @Override
+    public ActionDTO getActionByActionId(String actionType, String actionId, Integer tenantId)
+            throws ActionMgtException {
+
+        try {
+            ActionDTO actionDTO = actionManagementDAO.getActionByActionId(actionType, actionId, tenantId);
+            if (actionDTO == null) {
+                return null;
+            }
+
+            // Populate action properties
+            return getActionDTOWithPopulatedProperties(actionDTO, tenantId);
+        } catch (ActionMgtException | ActionPropertyResolverException e) {
+            throw ActionManagementExceptionHandler.handleServerException(
+                    ErrorMessage.ERROR_WHILE_RETRIEVING_ACTION_BY_ID, e);
+        }
+    }
+
+    @Override
+    public void updateAction(ActionDTO updatingActionDTO, ActionDTO existingActionDTO, Integer tenantId)
+            throws ActionMgtException {
+
+        NamedJdbcTemplate jdbcTemplate = new NamedJdbcTemplate(IdentityDatabaseUtil.getDataSource());
+        try {
+            jdbcTemplate.withTransaction(template -> {
+                ActionDTOBuilder updatingActionDTOBuilder = new ActionDTOBuilder(updatingActionDTO);
+                // Encrypt authentication secrets
+                encryptUpdatingAuthSecrets(updatingActionDTOBuilder, existingActionDTO);
+                // Resolve action properties
+                ActionDTO resolvedUpdatingActionDTO =
+                        getActionDTOWithResolvedUpdatingProperties(updatingActionDTOBuilder.build(), existingActionDTO,
+                                tenantId);
+
+                actionManagementDAO.updateAction(resolvedUpdatingActionDTO, existingActionDTO, tenantId);
+                return null;
+            });
+        } catch (TransactionException e) {
+            handleActionPropertyResolverClientException(e.getCause());
```

Comment:
This seems to be catching a TransactionException. In that case looks like the logic within this method will not be executed

Reply:
The withTransaction() method in jdbctemplate class wraps every exception with a TransactionException. 
If we want to check the exact cause we can check it using the e.getCause().

Wrote unit tests also to validate this scenario.
https://github.com/wso2/carbon-identity-framework/pull/6119/files#diff-b094e19b41079e4bf3e4f26aadb36b1de3e74044332b125c03222d2e8a462fbaR151-R166:~:text=testAddActionWithActionPropertyResolverClientException